### ISSUE TITLE
Cmake version tracking issue

### ISSUE THREAD

**Issue by dundargoc** [neovim/neovim#24004]
A tracking issue of pros and cons of upgrading the minimum cmake version.

## 3.11
- FindIconv module, meaning we can remove `cmake/FindIconv.cmake`
- FetchContent module, which is essentially "ExternalProject but good".
    - It is way more flexible and less hacky than ExternalProject. For example, it has a `FETCHCONTENT_FULLY_DISCONNECTED` variable users can set to disable download/update step which distributions can use, which would eliminate the `USE_EXISTING_SRC_DIR` option and the use-existing-src test in `build.yml`
    - TODO: Probably many other benefits, research more later.

## 3.12
- `CONFIGURE_DEPENDS` flag for `file(GLOB)`
    - We're already using these behind a version check.
## 3.13
- QOL `-B` and `-S` flags
    - Honestly these are mandatory IMO. I cringe every time I see someone do `mkdir build; cd build; cmake ..`.
## 3.14
- `FetchContent_MakeAvailable`
    - As I understand it, it basically reduces a bunch of boilerplate required when using the `FetchContent` module.
#### Drops support for
- Debian Buster
## 3.15
- `--install` flag
    - Possible to install neovim the same way across all platforms and generators.
- `ADDITIONAL_CLEAN_FILES` property
    - Provide a better cross-platform `clean` target.
- `$<REMOVE_DUPLICATES>` generator expression
    - Allows us to remove `cmake.config/versiondef_old.h.in`
## 3.16
- Didn't find anything useful
## 3.17
- Didn't find anything useful
#### Drops support for
- Ubuntu 20.04 
## 3.18
- CheckLinkerFlag module
- `REQUIRED` option for all "find_" commands.

## 3.19
- `string` gained `JSON` subcommands
    - Allows us to specify dependencies in JSON instead of our bespoke format in `cmake.deps/deps.txt`

#### Drops support for
- Debian Bullseye

**Comment by clason** [comment]
And to complement this excellent write-up, a summary of which distro version comes with which CMake version:

### Debian
* Buster (oldstable, LTS until June 2024): 3.13 (backports: 3.18)
* Bullseye (stable, LTS until June 2026): 3.18 (backports: 3.25)
* Bookworm (new stable): 3.25

### Ubuntu
* 20.04 (LTS until 2025): 3.16
* 22.04 (LTS until 2027): 3.22
* 22.10: 3.24
* 23.04: 3.25

**Comment by dundargoc** [comment]
I think I'll need to make a second and more thorough look at the cmake release notes, so consider the issue text as a work in progress for now. I should probably also include cmake versions for some other popular distros.

I kinda convinced myself with when writing this that we'd benefit a lot from `FetchContent`, so 3.11 and maybe 3.14 would be nice. I think that'd be the next biggest improvement we can strive for. I'd personally be satisfied with a bump to 3.13 as that'd give us `FetchContent` and `-B`/`-S` flags (I know I can already use this myself, but we can't do this in the Makefile or wiki for example).

**Comment by dundargoc** [comment]
I don't think we need to track backport/non-standard versions. I interpreted this issue to be about mainline/release distro versions only without any additional modifications. It is e.g. trivial to simply build latest cmake on any system.

**Comment by clason** [comment]
Yup, just added that for the sake of completeness (and to prevent "but did you know you could install later versions manually?" comments).
